### PR TITLE
Review H1 2023 changes

### PR DESCRIPTION
**Pull request by rube-de**

### CODE REVIEW

**Review by mathewmeconry — CHANGES_REQUESTED** [review]

File: src/multisig/MultiSigWalletMaster.sol [hunk 1]
```diff
@@ -164,14 +200,18 @@ function toBytes (uint256 x) public pure returns (bytes memory result) {
...
+      revert Multisig_InvalidDestination(destination);
+    }
     _setSigner(destination, signers[source]);
     _setSigner(source, 0);
   }
 
   function _setSigner(address signer, uint8 signaturesNeeded) private {
-    require(!Address.isContract(signer), "signer cannot be a contract");
-    require(signer != address(0x0), "0x0 signer");
+    if (Address.isContract(signer) || signer == address(0x0)) {
```
[the remark this answers is not in the document]
Comment (rube-de) [reply src/multisig/MultiSigWalletMaster.sol:212]: Our multisig should always have a assumed valid signer, this more like a safe guard beforehand, that you don't add a zero address signer, remove all other signers and lose access. It's debatable as you could set any other random address. As this is not a daily tx I would let it in.

File: scripts/migrate_empty_token.js [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+  const { deploy } = deployments;
+
+  const feeData = await ethers.provider.getFeeData();
+
+  // deploy shares contract
+  const symbol = "SHR";
+  const name = "Test Shares";
+  const terms = "test.ch/terms";
+  const totalShares = 10500000;
+  const recoveryHub = "0xc6AfD3b605957b9BA94143F59d451c657F507516";
```
[the remark this answers is not in the document]
Comment (rube-de) [reply scripts/migrate_empty_token.js:34]: that was a one time script to remove a name from a unused contract. deleted the script as it shouldn't be in the main branch.

File: deployments/mainnet/InvalidContract.json [hunk 3]
```diff
@@ -0,0 +1,797 @@
+{
+  "address": "0x27A02627F4D683594984984f0fD2dE9Aa3C8B2f5",
```
[the remark this answers is not in the document]
Comment (rube-de) [reply deployments/mainnet/InvalidContract.json:2]: it's an unused contract, was just deployed for name change to "Invalid Contract"

File: src/brokerbot/PaymentHub.sol [hunk 4]
```diff
@@ -179,119 +247,218 @@ contract PaymentHub {
     ///@dev This function needs to be called before using the PaymentHub the first time with a new ERC20 token.
     ///@param erc20In The erc20 addresse to approve.
     function approveERC20(address erc20In) external {
-        IERC20(erc20In).approve(address(uniswapRouter), 0x8000000000000000000000000000000000000000000000000000000000000000);
+        IERC20(erc20In).approve(address(uniswapRouter), type(uint256).max);
```
[the remark this answers is not in the document]
Comment (rube-de) [reply src/brokerbot/PaymentHub.sol:250]: type(uint256).max is > than 2*255, like this the if-block also doesn't get executed in the ERC20Flaggable.
The approval transaction will be slightly more expensive, but for wallet UX it's better to comply with the standard infinity approval like also USDT uses.

File: src/brokerbot/PaymentHub.sol [hunk 5]
```diff
@@ -179,119 +247,218 @@ contract PaymentHub {
...
-            uint256 proceeds = IBrokerbot(recipient).processIncoming(IERC20(token), address(this), amountToSell, ref);
-            IERC20 currency = IBrokerbot(recipient).base();
-            currency.transfer(msg.sender, exFee);
-            currency.transfer(seller, proceeds - exFee);
+        if (permitInfo.exFee > 0){
+            uint256 proceeds = _sellShares(brokerbot, shares, seller, address(this), amountToSell, ref);
+            IERC20 currency = brokerbot.base();
+            currency.transfer(msg.sender, permitInfo.exFee);
+            currency.transfer(recipient, proceeds - permitInfo.exFee);
+            return proceeds - permitInfo.exFee;
```
[the remark this answers is not in the document]
Comment (rube-de) [reply src/brokerbot/PaymentHub.sol]: fixed with safetransfer

**Pull request merged**